### ISSUE TITLE
The training criterion is not a number error for relu activation function

### ISSUE THREAD

**Issue by mbetser** [Microsoft/CNTK#385]
Hi guys,

First thanks for the good work. What I am trying to do: I have a list of time series which are counts of events over time and I want to detect a problem so my output is actually binary but I treat it as two classes because I am interested in the probability of the event.
I have tried 1d convolutional networks (1d convolution with 16 channels) on theano and was able train the models correctly. The model I have tried was close to the 02_Convolution.ndl example so I decided to start from there. The differences was that it was using dropouts and a relu dense layer instead of the sigmoid layer and that the convolutions were 1d.

With cntk, training is crashing at the very beginning with the error (log attached):
"The training criterion is not a number (NAN). Stop"
I have tried to train without dropouts and sigmoid activation, and this configuration did not crash but was not converging either.

I am not sure what I am doing wrong at that point. The major difference with the mnist case, is that I am using a 1d convolution instead of 2d and maybe I am not configuring the network correctly?
One weird thing I have noticed in the log is that the feature input has this shape:
 features[1 x 64 x 16 x \* {W=64, H=16, C=1}]
what bothers me is that the height is 16 instead of the channel. But I am sure I have defined the channel to be 16 in the ndl file (also attached). The image layout is cudnn.
Also there is no mention of the imageLayout parameter for the ImageInput function in the cntk book even though it is used in the examples

Thanks for your help

Michael

[myConvolution.txt](https://github.com/Microsoft/CNTK/files/230665/myConvolution.txt)
[log.txt](https://github.com/Microsoft/CNTK/files/230698/log.txt)

**Comment by Alexey-Kamenev** [comment]
Hello,
Can you please verify that you set hyper-parameters like learning rate and momentum to the same values as in your Theano experiments? For example, we have `learningRatesPerSample` as well as `learningRatesPerMB` so depending on what Theano uses, the end result (i.e. effective learning rate per sample) could be different.

**Comment by mbetser** [comment]
Hi Alexey,

I forgot to mention that I am using the keras framework on top of theano. It seems that the learning rate there is per mini batch.
So I have tried to use different learning rates using the learningRatesPerMB variable from 100.0 to 0.00001 but I still get the same NAN error for relu networks.
How can the training criterion be a NAN? I am using the crossentropywithsoftmax criterion so there is a log in there which could throw a nan without an espilon guard. Any advice on what to try next?

Thanks,

Michael.

**Comment by Alexey-Kamenev** [comment]
Hi Michael,
Can you please try using our most recent binary release (or, if possible, latest sources from master)?

**Comment by Alexey-Kamenev** [comment]
Also, what is the range of your input data? Another thing to look for is parameter randomization: make sure you are using the same values as in your keras setup (look at convWScale and convBValue in your NDL)

**Comment by mbetser** [comment]
It turned out the problem was a mix of 
- not using the right momentum and learning rate parameter (Mini batch was the good one in my case)
- not using the normalization parameter function I was using in theano (I really though I was)
- not cleaning the cntk model files from previous sessions (which is not done by default. Need to add "makeMode = false" to the train section)

Thanks for the help I converges now with very close performances.
One more question though. In theano there is way to define dropout per layer instead of one for the whole network. Is it possible to do that with cntk?

**Comment by Alexey-Kamenev** [comment]
Thanks for letting us know! In fact, this is a common source of issues when migrating between the toolkits. We've started documenting some specific differences which hopefully will help people to avoid such issues in the future.
As for the dropout: it is currently a global parameter so there no convenient way to set it per-layer. You can, however, change it on per-epoch basis.
